File: web/src/lib/judge-activity.ts
/**
 * Everything the judge is being asked to do, in one list.
 *
 * Three different things in this system end up as a job on the judge, and until
 * now each was only visible from its own page:
 *
 *   a code submission  -> one `judgement` row per attempt (a rejudge is a new
 *                         request, so superseded rows are shown too)
 *   a hack             -> one `p1HackAttempt` row; the participant sends an
 *                         input, the judge runs someone else's code on it
 *   a validator run    -> one `p1Answer` row, scored at section close
 *
 * When the judge is slow or a container has fallen over, the question is never
 * "how are submissions doing" — it is "what is it chewing on", and the answer
 * has to include the hacks. So they are normalised into one shape here and
 * merged, with whatever is still in flight first: that is the part that can
 * still go wrong.
 *
 * Nothing here is participant-facing. It carries jury detail, verdicts on
 * hacks, and source code, and is only ever reached through an admin or
 * evaluator session.
 */
import { and, desc, eq, isNotNull, sql } from "drizzle-orm";

import { db } from "@/db";
import { user } from "@/db/auth-schema";
import {
  judgement,
  p1Answer,
  p1HackAttempt,
  p1HackQuestion,
  p1Question,
  question,
  submission,
} from "@/db/schema";

import { errors } from "./api";
import { judge } from "./judge";

export type WorkKind = "submission" | "hack" | "validator";

/** One request to the judge, whatever kind it is. */
export type Work = {
  kind: WorkKind;
  /** Stable across refreshes, and what the row is keyed and linked by. */
  key: string;
  /** The row id in its own table — "hack/12" addresses the detail endpoint. */
  ref: string;
  job_id: string | null;
  /** pending = not yet accepted by the judge. error = it answered, badly. */
  state: "pending" | "queued" | "running" | "done" | "error";
  live: boolean;
  participant_id: string;
  name: string;
  /** What is being judged, as a person would name it. */
  target: string;
  /** The package id actually sent to the judge. */
  problem_id: string | null;
  language: string | null;
  progress: { done: number; total: number } | null;
  /** The short result: a verdict code, or a hack outcome. */
  verdict: string | null;
  /**
   * How it finished, as a chip. Submissions carry a verdict and the table
   * spells that out itself; a hack or a validator run has no verdict, so the
   * word and its colour are decided here — otherwise every one of them renders
   * the same grey dot and "broke the solution" looks like "rejected".
   */
  label: string | null;
  tone: "success" | "destructive" | "warning" | "info" | "neutral";
  /** The rest of the sentence, after the chip. */
  outcome: string | null;
  retries: number;
  attempt: number | null;
  superseded: boolean;
  created_at: string;
  ended_at: string | null;
  duration_ms: number | null;
};

const LIVE = new Set(["pending", "queued", "running"]);

function shape(w: Omit<Work, "live" | "duration_ms">): Work {
  const started = new Date(w.created_at).getTime();
  const ended = w.ended_at ? new Date(w.ended_at).getTime() : null;
  return { ...w, live: LIVE.has(w.state), duration_ms: ended ? ended - started : null };
}

/**
 * The feed. `limit` caps each source, not the total, so a burst of submissions
 * cannot push every hack off the end of the list.
 */
export async function judgeActivity(limit = 200): Promise<Work[]> {
  const [subs, hacks, validators] = await Promise.all([
    db
      .select({ j: judgement, s: submission, name: user.name, title: question.title })
      .from(judgement)
      .innerJoin(submission, eq(submission.id, judgement.submissionId))
      .innerJoin(user, eq(user.id, submission.participantId))
      .innerJoin(question, eq(question.id, submission.questionId))
      .orderBy(desc(judgement.id))
      .limit(limit),
    db
      .select({ a: p1HackAttempt, q: p1HackQuestion, name: user.name })
      .from(p1HackAttempt)
      .innerJoin(p1HackQuestion, eq(p1HackQuestion.id, p1HackAttempt.questionId))
      .innerJoin(user, eq(user.id, p1HackAttempt.participantId))
      .orderBy(desc(p1HackAttempt.id))
      .limit(limit),
    db
      .select({ a: p1Answer, q: p1Question, name: user.name })
      .from(p1Answer)
      .innerJoin(p1Question, eq(p1Question.id, p1Answer.questionId))
      .innerJoin(user, eq(user.id, p1Answer.participantId))
      .where(isNotNull(p1Answer.scoreState))
      .orderBy(desc(p1Answer.updatedAt))
      .limit(limit),
  ]);

  const rows: Work[] = [
    ...subs.map(({ j, s, name, title }) =>
      shape({
        kind: "submission",
        key: `submission:${j.id}`,
        ref: `submission/${j.id}`,
        job_id: j.jobId,
        state: j.state === "done" && j.verdict === "IE" ? "error" : (j.state as Work["state"]),
        participant_id: s.participantId,
        name: name ?? s.participantId,
        target: title,
        problem_id: s.questionId,
        language: s.language,
        progress: j.progressTotal > 0 ? { done: j.progressDone, total: j.progressTotal } : null,
        verdict: j.verdict,
        label: null,
        tone: j.verdict === "AC" ? "success" : j.verdict ? "destructive" : "neutral",
        outcome: j.cancelled ? "Cancelled" : (j.message ?? null),
        retries: j.retries,
        attempt: j.attempt,
        superseded: Boolean(j.supersededAt),
        created_at: j.createdAt.toISOString(),
        ended_at: j.endedAt?.toISOString() ?? null,
      }),
    ),
    ...hacks.map(({ a, q, name }) =>
      shape({
        kind: "hack",
        key: `hack:${a.id}`,
        ref: `hack/${a.id}`,
        job_id: a.jobId,
        state: a.state === "done" && a.verdict === "IE" ? "error" : a.state,
        participant_id: a.participantId,
        name: name ?? a.participantId,
        target: q.title,
        problem_id: q.problemId,
        language: q.givenLanguage,
        progress: null,
        verdict: a.verdict,
        ...hackChip(a),
        retries: a.retries,
        attempt: null,
        superseded: false,
        created_at: a.createdAt.toISOString(),
        ended_at: a.endedAt?.toISOString() ?? null,
      }),
    ),
    ...validators.map(({ a, q, name }) =>
      shape({
        kind: "validator",
        key: `validator:${a.participantId}:${a.questionId}`,
        ref: `validator/${a.participantId}:${a.questionId}`,
        job_id: a.scoreJobId,
        state: (a.scoreState ?? "pending") as Work["state"],
        participant_id: a.participantId,
        name: name ?? a.participantId,
        target: q.title,
        problem_id: null,
        language: "python",
        progress: null,
        verdict: a.scoreState === "error" ? "IE" : null,
        label: a.scoreState === "error" ? "Judge error" : a.scoreState === "done" ? `Scored ${a.autoScore ?? 0}` : null,
        tone: a.scoreState === "error" ? "destructive" : "success",
        outcome: a.scoreState === "error" ? (a.scoreError ?? "the validator did not finish") : null,
        retries: 0,
        attempt: null,
        superseded: false,
        created_at: a.updatedAt.toISOString(),
        /* A p1Answer keeps one timestamp, not two, so there is no honest
         * duration to report for a validator run. Better an em dash than a
         * confident "0 ms". */
        ended_at: null,
      }),
    ),
  ];

  // In flight first, oldest at the top — the one that has been waiting longest
  // is the one worth looking at. Everything settled follows, newest first.
  return rows.sort((a, b) => {
    if (a.live !== b.live) return a.live ? -1 : 1;
    const t = new Date(a.created_at).getTime() - new Date(b.created_at).getTime();
    return a.live ? t : -t;
  });
}

type Chip = { label: string | null; tone: Work["tone"]; outcome: string | null };

function hackChip(a: typeof p1HackAttempt.$inferSelect): Chip {
  if (a.state !== "done") return { label: null, tone: "info", outcome: null };
  if (a.verdict === "IE") return { label: "Judge error", tone: "destructive", outcome: "the judge could not run it" };
  if (a.validInput === false) return { label: "Rejected", tone: "warning", outcome: a.invalidReason ?? "the input broke the constraints" };
  if (a.hacked === true) return { label: "Broke it", tone: "success", outcome: "the solution failed on this input" };
  if (a.hacked === false) return { label: "Held", tone: "neutral", outcome: "the solution survived" };
  return { label: null, tone: "neutral", outcome: null };
}

/** The same words, as one sentence, for the drawer. */
function hackOutcome(a: typeof p1HackAttempt.$inferSelect): string | null {
  const c = hackChip(a);
  return c.label ? (c.outcome ? `${c.label} — ${c.outcome}` : c.label) : null;
}

/** Counts for the header tiles, taken from the same rows the list is built from. */
export function summarise(rows: Work[]) {
  const live = rows.filter((r) => r.live);
  return {
    live: live.length,
    queued: live.filter((r) => r.state === "queued" || r.state === "pending").length,
    running: live.filter((r) => r.state === "running").length,
    retrying: live.filter((r) => r.retries > 0).length,
    errors: rows.filter((r) => r.state === "error" && !r.superseded).length,
    hacks: live.filter((r) => r.kind === "hack").length,
  };
}

// ---------------------------------------------------------------------------
// One request, in full
// ---------------------------------------------------------------------------

export type WorkDetail = Awaited<ReturnType<typeof workDetail>>;

/**
 * Everything known about one judge request, including what was sent to it.
 *
 * The three kinds do not share a shape, so this returns a tagged union rather
 * than a lowest common denominator: a hack has an input and no test counts, a
 * submission has test counts and no input.
 */
export async function workDetail(kind: string, id: string) {
  if (kind === "submission") return submissionDetail(Number(id));
  if (kind === "hack") return hackDetail(Number(id));
  if (kind === "validator") return validatorDetail(id);
  throw errors.notFound("job");
}

async function submissionDetail(judgementId: number) {
  const [row] = await db
    .select({ j: judgement, s: submission, name: user.name, title: question.title })
    .from(judgement)
    .innerJoin(submission, eq(submission.id, judgement.submissionId))
    .innerJoin(user, eq(user.id, submission.participantId))
    .innerJoin(question, eq(question.id, submission.questionId))
    .where(eq(judgement.id, judgementId));
  if (!row) throw errors.notFound("job");
  const { j, s } = row;

  // The testcase that failed, read back at the exact version it was judged
  // against — a package republished since would otherwise show the wrong data.
  const failing =
    j.firstFail !== null && j.problemVersion
      ? await judge.testcase(s.questionId, j.firstFail, j.problemVersion).catch(() => null)
      : null;

  const attempts = await db
    .select({ id: judgement.id, attempt: judgement.attempt, verdict: judgement.verdict, state: judgement.state, createdAt: judgement.createdAt })
    .from(judgement)
    .where(eq(judgement.submissionId, s.id))
    .orderBy(desc(judgement.attempt));

  return {
    kind: "submission" as const,
    who: { id: s.participantId, name: row.name ?? s.participantId },
    submitted_at: s.createdAt.toISOString(),
    target: { title: row.title, problem_id: s.questionId },
    source: s.source,
    language: s.language,
    request: {
      job_id: j.jobId,
      state: j.state,
      attempt: j.attempt,
      retries: j.retries,
      cancelled: j.cancelled,
      superseded_at: j.supersededAt?.toISOString() ?? null,
      problem_version: j.problemVersion,
      created_at: j.createdAt.toISOString(),
      ended_at: j.endedAt?.toISOString() ?? null,
    },
    result: {
      verdict: j.verdict,
      passed: j.passed,
      total: j.total,
      first_fail: j.firstFail,
      max_time_ms: j.maxTimeMs,
      max_memory_kb: j.maxMemoryKb,
      message: j.message,
      jury_detail: j.juryDetail,
      compile_output: j.compileOutput,
      progress: { done: j.progressDone, total: j.progressTotal },
    },
    failing_testcase: failing,
    attempts: attempts.map((a) => ({ ...a, createdAt: a.createdAt.toISOString() })),
  };
}

async function hackDetail(attemptId: number) {
  const [row] = await db
    .select({ a: p1HackAttempt, q: p1HackQuestion, name: user.name })
    .from(p1HackAttempt)
    .innerJoin(p1HackQuestion, eq(p1HackQuestion.id, p1HackAttempt.questionId))
    .innerJoin(user, eq(user.id, p1HackAttempt.participantId))
    .where(eq(p1HackAttempt.id, attemptId));
  if (!row) throw errors.notFound("job");
  const { a, q } = row;

  return {
    kind: "hack" as const,
    who: { id: a.participantId, name: row.name ?? a.participantId },
    submitted_at: a.createdAt.toISOString(),
    target: { title: q.title, problem_id: q.problemId },
    /* A hack sends an input, not code. The code is the solution under attack,
     * which every participant can already read on the question page. */
    input: a.input,
    source: q.givenSource,
    language: q.givenLanguage,
    request: {
      job_id: a.jobId,
      state: a.state,
      attempt: null,
      retries: a.retries,
      cancelled: false,
      superseded_at: null,
      problem_version: null,
      created_at: a.createdAt.toISOString(),
      ended_at: a.endedAt?.toISOString() ?? null,
    },
    result: {
      valid_input: a.validInput,
      invalid_reason: a.invalidReason,
      hacked: a.hacked,
      verdict: a.verdict,
      points_awarded: a.pointsAwarded,
      outcome: hackOutcome(a),
    },
    stakes: { hack_points: q.hackPoints, fail_penalty: q.failPenalty },
  };
}

async function validatorDetail(ref: string) {
  const [participantId, questionId] = ref.split(":");
  const [row] = await db
    .select({ a: p1Answer, q: p1Question, name: user.name })
    .from(p1Answer)
    .innerJoin(p1Question, eq(p1Question.id, p1Answer.questionId))
    .innerJoin(user, eq(user.id, p1Answer.participantId))
    .where(and(eq(p1Answer.participantId, participantId), eq(p1Answer.questionId, Number(questionId))));
  if (!row) throw errors.notFound("job");
  const { a, q } = row;
  const entries = Array.isArray(a.answer) ? (a.answer as unknown[]).map(String) : a.answer == null ? [] : [String(a.answer)];

  return {
    kind: "validator" as const,
    who: { id: a.participantId, name: row.name ?? a.participantId },
    submitted_at: a.updatedAt.toISOString(),
    target: { title: q.title, problem_id: null },
    /* The validator itself is an answer key. It is not returned, even here —
     * this page is read over shoulders during a contest. */
    entries,
    language: "python",
    request: {
      job_id: a.scoreJobId,
      state: a.scoreState ?? "pending",
      attempt: null,
      retries: 0,
      cancelled: false,
      superseded_at: null,
      problem_version: null,
      created_at: a.updatedAt.toISOString(),
      /* One timestamp on the row, so there is no honest duration. */
      ended_at: null,
    },
    result: {
      score: a.autoScore,
      points_per_entry: q.pointsPerEntry,
      error: a.scoreError,
    },
  };
}

export const _sql = sql;

File: engine/engine/admin/judge_activity.py
"""Everything the judge is being asked to do, in one list, for staff.

Four things become judge jobs: a code submission (a judgement row per
attempt), a practice run (a practice_run row), a hack (a p1_hack_attempt row),
and a validator run (a p1_answer row, scored at section close). They are
normalised into one shape and merged, with whatever is still in flight first
-- that is the part that can still go wrong.
"""

from __future__ import annotations

from typing import Any

import sqlalchemy as sa

from engine.admin.overview import health
from engine.admin.records import failing_testcase
from engine.core import clock, db, errors
from engine.schema import (
    judgement,
    p1_answer,
    p1_hack_attempt,
    p1_hack_question,
    p1_hack_solution,
    p1_question,
    practice_run,
    question,
    submission,
    user,
)

LIVE = ("pending", "queued", "running")


def activity(limit: int = 200) -> list[dict[str, Any]]:
    """Every judge job, live ones oldest first, then finished ones newest first.

    Each source is capped separately, so a burst of submissions cannot push every
    hack off the list.
    """
    with db.transaction() as conn:
        work = (
            _submissions(conn, limit)
            + _runs(conn, limit)
            + _hacks(conn, limit)
            + _validators(conn, limit)
        )
    live = [w for w in work if w["live"]]
    done = [w for w in work if not w["live"]]
    live.sort(key=lambda w: w["created_at"])
    done.sort(key=lambda w: w["created_at"], reverse=True)
    return live + done


def overview() -> dict[str, Any]:
    """Judge health and its work in one response.

    Sent together so the tiles and the list never disagree about whether the judge is up.
    """
    status = health()
    work = activity()
    return {
        "judge": status["judge"],
        "backlog": status["backlog"],
        "work": work,
        "counts": summarise(work),
        "server_now": clock.now_ms(),
    }


def summarise(work: list[dict[str, Any]]) -> dict[str, int]:
    live = [w for w in work if w["live"]]
    return {
        "live": len(live),
        "queued": sum(1 for w in live if w["state"] in ("queued", "pending")),
        "running": sum(1 for w in live if w["state"] == "running"),
        "retrying": sum(1 for w in live if w["retries"] > 0),
        "errors": sum(1 for w in work if w["state"] == "error" and not w["superseded"]),
        "hacks": sum(1 for w in live if w["kind"] == "hack"),
    }


def _shape(**w: Any) -> dict[str, Any]:
    """Finish one job row: whether it is live, how long it took, and ISO timestamps.

    A job the judge finished with IE is shown as an error: it is ours to fix.
    """
    if w["state"] == "done" and w["verdict"] == "IE":
        w["state"] = "error"
    started, ended = w["created_at"], w["ended_at"]
    w["live"] = w["state"] in LIVE
    w["duration_ms"] = clock.ms(ended) - clock.ms(started) if ended else None
    w["created_at"], w["ended_at"] = clock.iso(started), clock.iso(ended)
    return w


def _verdict_tone(verdict: str | None) -> str:
    if verdict == "AC":
        return "success"
    return "destructive" if verdict else "neutral"


def _submissions(conn: sa.Connection, limit: int) -> list[dict[str, Any]]:
    found = conn.execute(
        sa.select(
            judgement,
            submission.c.participant_id,
            submission.c.question_id,
            submission.c.language,
            user.c.name,
            question.c.title,
        )
        .join(submission, submission.c.id == judgement.c.submission_id)
        .join(user, user.c.id == submission.c.participant_id)
        .join(question, question.c.id == submission.c.question_id)
        .order_by(judgement.c.id.desc())
        .limit(limit)
    ).all()
    out = []
    for j in found:
        progress = None
        if j.progress_total > 0:
            progress = {"done": j.progress_done, "total": j.progress_total}
        row = _shape(
            kind="submission",
            key=f"submission:{j.id}",
            ref=f"submission/{j.id}",
            job_id=j.job_id,
            state=j.state,
            participant_id=j.participant_id,
            name=j.name or j.participant_id,
            target=j.title,
            problem_id=j.question_id,
            question_id=j.question_id,
            language=j.language,
            progress=progress,
            verdict=j.verdict,
            label=None,
            tone=_verdict_tone(j.verdict),
            outcome="Cancelled" if j.cancelled else j.message,
            retries=j.retries,
            attempt=j.attempt,
            superseded=j.superseded_at is not None,
            created_at=j.created_at,
            ended_at=j.ended_at,
        )
        out.append(row)
    return out


def _runs(conn: sa.Connection, limit: int) -> list[dict[str, Any]]:
    """Practice runs keep no source and no output; the row is all there is to show."""
    found = conn.execute(
        sa.select(practice_run, user.c.name, question.c.title)
        .join(user, user.c.id == practice_run.c.participant_id)
        .join(question, question.c.id == practice_run.c.question_id)
        .order_by(practice_run.c.id.desc())
        .limit(limit)
    ).all()
    return [
        _shape(
            kind="run",
            key=f"run:{r.id}",
            ref=None,
            job_id=r.job_id,
            state=r.state,
            participant_id=r.participant_id,
            name=r.name or r.participant_id,
            target=r.title,
            problem_id=r.question_id,
            question_id=r.question_id,
            language=r.language,
            progress=None,
            verdict=r.verdict,
            label=None,
            tone=_verdict_tone(r.verdict),
            outcome=r.message,
            retries=0,
            attempt=None,
            superseded=False,
            created_at=r.created_at,
            ended_at=r.ended_at,
        )
        for r in found
    ]


def _hacks(conn: sa.Connection, limit: int) -> list[dict[str, Any]]:
    found = conn.execute(
        sa.select(
            p1_hack_attempt,
            p1_hack_question.c.title,
            p1_hack_question.c.problem_id,
            p1_hack_solution.c.language,
            user.c.name,
        )
        .join(p1_hack_question, p1_hack_question.c.id == p1_hack_attempt.c.question_id)
        .outerjoin(p1_hack_solution, p1_hack_solution.c.id == p1_hack_attempt.c.solution_id)
        .join(user, user.c.id == p1_hack_attempt.c.participant_id)
        .order_by(p1_hack_attempt.c.id.desc())
        .limit(limit)
    ).all()
    return [
        _shape(
            kind="hack",
            key=f"hack:{a.id}",
            ref=f"hack/{a.id}",
            job_id=a.job_id,
            state=a.state,
            participant_id=a.participant_id,
            name=a.name or a.participant_id,
            target=a.title,
            problem_id=a.problem_id,
            question_id=a.question_id,
            language=a.language,
            progress=None,
            verdict=a.verdict,
            **hack_chip(a),
            retries=a.retries,
            attempt=None,
            superseded=False,
            created_at=a.created_at,
            ended_at=a.ended_at,
        )
        for a in found
    ]


def _validators(conn: sa.Connection, limit: int) -> list[dict[str, Any]]:
    found = conn.execute(
        sa.select(p1_answer, p1_question.c.title, user.c.name)
        .join(p1_question, p1_question.c.id == p1_answer.c.question_id)
        .join(user, user.c.id == p1_answer.c.participant_id)
        .where(p1_answer.c.score_state.is_not(None))
        .order_by(p1_answer.c.updated_at.desc())
        .limit(limit)
    ).all()
    out = []
    for a in found:
        verdict = None
        label = None
        tone = "success"
        outcome = None
        if a.score_state == "error":
            verdict = "IE"
            label = "Judge error"
            tone = "destructive"
            outcome = a.score_error or "the validator did not finish"
        elif a.score_state == "done":
            label = f"Scored {a.auto_score or 0}"
        row = _shape(
            kind="validator",
            key=f"validator:{a.participant_id}:{a.question_id}",
            ref=f"validator/{a.participant_id}:{a.question_id}",
            job_id=a.score_job_id,
            state=a.score_state or "pending",
            participant_id=a.participant_id,
            name=a.name or a.participant_id,
            target=a.title,
            problem_id=None,
            question_id=a.question_id,
            language="python",
            progress=None,
            verdict=verdict,
            label=label,
            tone=tone,
            outcome=outcome,
            retries=0,
            attempt=None,
            superseded=False,
            # One timestamp on the row, so there is no honest duration to report.
            created_at=a.updated_at,
            ended_at=None,
        )
        out.append(row)
    return out


def hack_chip(a: sa.Row) -> dict[str, Any]:
    """How a hack finished, as a label and a colour.

    A hack has no verdict for the table to spell out.
    """
    if a.state != "done":
        return {"label": None, "tone": "info", "outcome": None}
    if a.verdict == "IE":
        return {
            "label": "Judge error",
            "tone": "destructive",
            "outcome": "the judge could not run it",
        }
    if a.valid_input is False:
        return {
            "label": "Rejected",
            "tone": "warning",
            "outcome": a.invalid_reason or "the input broke the constraints",
        }
    if a.hacked is True:
        return {
            "label": "Broke it",
            "tone": "success",
            "outcome": "the solution failed on this input",
        }
    if a.hacked is False:
        return {"label": "Held", "tone": "neutral", "outcome": "the solution survived"}
    return {"label": None, "tone": "neutral", "outcome": None}


def detail(kind: str, ref: str) -> dict[str, Any]:
    """One job in full. `kind` and `ref` are the two halves of its activity row's `ref`."""
    if kind == "submission":
        return _submission_detail(int(ref))
    if kind == "hack":
        return _hack_detail(int(ref))
    if kind == "validator":
        return _validator_detail(ref)
    raise errors.not_found("job")


def _request(
    job_id: str | None, state: str, created_at: Any, ended_at: Any, **extra: Any
) -> dict[str, Any]:
    """What was asked of the judge, in one shape for every kind of job."""
    base = {
        "job_id": job_id,
        "state": state,
        "attempt": None,
        "retries": 0,
        "cancelled": False,
        "superseded_at": None,
        "problem_version": None,
        "created_at": clock.iso(created_at),
        "ended_at": clock.iso(ended_at),
    }
    return base | extra


def _submission_detail(judgement_id: int) -> dict[str, Any]:
    with db.transaction() as conn:
        j = conn.execute(
            sa.select(
                judgement,
                submission.c.participant_id,
                submission.c.question_id,
                submission.c.source,
                submission.c.language,
                submission.c.created_at.label("submitted_at"),
                user.c.name,
                question.c.title,
            )
            .join(submission, submission.c.id == judgement.c.submission_id)
            .join(user, user.c.id == submission.c.participant_id)
            .join(question, question.c.id == submission.c.question_id)
            .where(judgement.c.id == judgement_id)
        ).one_or_none()
        if j is None:
            raise errors.not_found("job")
        attempts = _attempts(conn, j.submission_id)
    return {
        "kind": "submission",
        "who": {"id": j.participant_id, "name": j.name or j.participant_id},
        "submitted_at": clock.iso(j.submitted_at),
        "target": {"title": j.title, "problem_id": j.question_id, "question_id": j.question_id},
        "source": j.source,
        "language": j.language,
        "request": _request(
            j.job_id,
            j.state,
            j.created_at,
            j.ended_at,
            attempt=j.attempt,
            retries=j.retries,
            cancelled=j.cancelled,
            superseded_at=clock.iso(j.superseded_at),
            problem_version=j.problem_version,
        ),
        "result": {
            "verdict": j.verdict,
            "passed": j.passed,
            "total": j.total,
            "first_fail": j.first_fail,
            "max_time_ms": j.max_time_ms,
            "max_memory_kb": j.max_memory_kb,
            "message": j.message,
            "jury_detail": j.jury_detail,
            "compile_output": j.compile_output,
            "progress": {"done": j.progress_done, "total": j.progress_total},
        },
        "failing_testcase": failing_testcase(j.question_id, j.first_fail, j.problem_version),
        "attempts": attempts,
    }


def _attempts(conn: sa.Connection, submission_id: int) -> list[dict[str, Any]]:
    """Every judgement of one submission, newest attempt first -- the rejudge history."""
    found = conn.execute(
        sa.select(
            judgement.c.id,
            judgement.c.attempt,
            judgement.c.verdict,
            judgement.c.state,
            judgement.c.created_at,
        )
        .where(judgement.c.submission_id == submission_id)
        .order_by(judgement.c.attempt.desc())
    ).all()
    return [
        {
            "id": a.id,
            "attempt": a.attempt,
            "verdict": a.verdict,
            "state": a.state,
            "createdAt": clock.iso(a.created_at),
        }
        for a in found
    ]


def _hack_detail(attempt_id: int) -> dict[str, Any]:
    with db.transaction() as conn:
        a = conn.execute(
            sa.select(
                p1_hack_attempt,
                p1_hack_question.c.title,
                p1_hack_question.c.problem_id,
                p1_hack_solution.c.source,
                p1_hack_solution.c.language,
                p1_hack_question.c.hack_points,
                p1_hack_question.c.fail_penalty,
                user.c.name,
            )
            .join(p1_hack_question, p1_hack_question.c.id == p1_hack_attempt.c.question_id)
            .outerjoin(p1_hack_solution, p1_hack_solution.c.id == p1_hack_attempt.c.solution_id)
            .join(user, user.c.id == p1_hack_attempt.c.participant_id)
            .where(p1_hack_attempt.c.id == attempt_id)
        ).one_or_none()
    if a is None:
        raise errors.not_found("job")
    chip = hack_chip(a)
    if not chip["label"]:
        outcome = None
    elif chip["outcome"]:
        outcome = f"{chip['label']} — {chip['outcome']}"
    else:
        outcome = chip["label"]
    return {
        "kind": "hack",
        "who": {"id": a.participant_id, "name": a.name or a.participant_id},
        "submitted_at": clock.iso(a.created_at),
        "target": {"title": a.title, "problem_id": a.problem_id, "question_id": a.question_id},
        # A hack sends an input, not code; the code is the copy under attack,
        # which is gone if an organiser removed that language afterwards.
        "input": a.input,
        "source": a.source or "",
        "language": a.language,
        "request": _request(a.job_id, a.state, a.created_at, a.ended_at, retries=a.retries),
        "result": {
            "valid_input": a.valid_input,
            "invalid_reason": a.invalid_reason,
            "hacked": a.hacked,
            "verdict": a.verdict,
            "points_awarded": a.points_awarded,
            "outcome": outcome,
        },
        "stakes": {"hack_points": a.hack_points, "fail_penalty": a.fail_penalty},
    }


def _validator_detail(ref: str) -> dict[str, Any]:
    participant_id, _, question_id = ref.partition(":")
    if not question_id.isdigit():
        raise errors.not_found("job")
    with db.transaction() as conn:
        a = conn.execute(
            sa.select(p1_answer, p1_question.c.title, p1_question.c.points_per_entry, user.c.name)
            .join(p1_question, p1_question.c.id == p1_answer.c.question_id)
            .join(user, user.c.id == p1_answer.c.participant_id)
            .where(
                p1_answer.c.participant_id == participant_id,
                p1_answer.c.question_id == int(question_id),
            )
        ).one_or_none()
    if a is None:
        raise errors.not_found("job")
    if isinstance(a.answer, list):
        entries = [str(e) for e in a.answer]
    elif a.answer is None:
        entries = []
    else:
        entries = [str(a.answer)]
    return {
        "kind": "validator",
        "who": {"id": a.participant_id, "name": a.name or a.participant_id},
        "submitted_at": clock.iso(a.updated_at),
        "target": {"title": a.title, "problem_id": None, "question_id": a.question_id},
        # The validator is an answer key: not returned, even here.
        "entries": entries,
        "language": "python",
        "request": _request(a.score_job_id, a.score_state or "pending", a.updated_at, None),
        "result": {
            "score": a.auto_score,
            "points_per_entry": a.points_per_entry,
            "error": a.score_error,
        },
    }
